apps/labs/posts/libsf-error-state.md

```diff
@@ -0,0 +1,744 @@
+involving [Generalized universal functions](https://numpy.org/doc/stable/reference/c-api/generalized-ufuncs.html)
+(gufuncs for short) so he wrote new machinery from scratch. Ufuncs and gufuncs created with the new machinery live in a
+separate extension module from those created with the old machinery []. It will be a great win in terms of simplifying
+`scipy.special`'s build process when all ufuncs can be moved over to the new infrastructure but in the short term there
+was a problem.
+
+## Error handling redux
+
+After splitting them between multiple extension modules, the error policy state could no longer be easily shared
+between ufuncs. We noticed something was wrong due to a failure in a doctest for
+[`special.seterr`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.special.seterr.html#scipy.special.seterr).
+The expected error was not being raised.
+
+```
+    Examples
+    --------
+    >>> import scipy.special as sc
+    >>> from pytest import raises
+    >>> sc.gammaln(0)
+    inf
+    >>> olderr = sc.seterr(singular='raise')
+    >>> with raises(sc.SpecialFunctionError):
+    ...     sc.gammaln(0)
+    ...
+    >>> _ = sc.seterr(**olderr)
+```
+
+[`gammaln`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.special.gammaln.html) was one of a handful of
+ufuncs moved to the new extension module.
+
+Both extension modules contained a separate copy of the state for managing error handling policies, but the user facing
+`special.errstate` could only see and change the state from the first extension module. While investigating, we also
+found that, for some reason, there was a separate extension module just for the function
+[`ellip_harm_2`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.special.ellip_harm_2.html#scipy.special.ellip_harm_2).
+As expected, `special.errstate` did not and had never worked for `ellip_harm_2` either, but since there were no relevant
+tests, no one knew.
+
+We saw three options:
+
+1. Have the Python interface for modifying the error handling state (`special.errstate` and
+   [`special.seterr`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.special.seterr.html)) update the state
+   in each extension module, taking care that the state remains synchronized.
+
+2. Extract the error handling state and primitives for managing it into a shared library that would be linked with
+   each extension module.
+
+3. Keep the error handling state in one of the extension modules and have it retrieved from there by the others.
+
+We chose to create a shared library because it seemed like the more principled option []. As the saying goes: we do
+things not because they are easy, but because we think they will be easy []. Quansight Labs Co-director Ralf Gommers
+(rgommers) knew from long experience that this may be difficult to get right, but trusted that we'd be able to figure
```

Comment:
several more empty brackets in this paragraph. also make "rgommers" a GitHub handle link?